### ISSUE TITLE
Parsing Multipart EML fails

### ISSUE THREAD

**Issue by frereit** [cirosec-studis/python-zimbra-web#59]
This fails sometimes, because we're changing `dict_mail['body']` to a string while looping over it.

https://github.com/cirosec-studis/python-zimbra-web/blob/9df3f5eb8cc61729cb7642ca3b4478ef619f22c0/src/zimbraweb/emlparsing.py#L47-L50

# Reproducing

Download and extract [email.zip](https://github.com/cirosec-studis/python-zimbra-web/files/7513666/email.zip)

```python
>>> from zimbraweb import emlparsing, ZimbraUser
>>> with open("email.eml") as f:
...     raw = f.read()
... 
>>> user = ZimbraUser("https://studgate.dhbw-mannheim.de")
>>> user.login("s000000", "#######")
True
>>> emlparsing.parse_eml(user, raw)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "F:\Coding\python\SMTP-Bridge\smtp-bridge\env\lib\site-packages\zimbraweb\emlparsing.py", line 48, in parse_eml
    if b.get('Content-Type')[:b.get('Content-Type').find(";")] == "text/plain":
AttributeError: 'NoneType' object has no attribute 'find'
```

# Fixing this

This bug is my fault, we do need a different variable and then just assign after the loop.

**Comment by JM-Lemmi** [comment]
Aah, there was a reason I did the extra variable, i just couldn't remember why I did it when you commented on the PR 😂

**Comment by frereit** [comment]
This is not fixed, the problem has to be something different because the error message didn't even change.

**Comment by JM-Lemmi** [comment]
Shit

**Comment by JM-Lemmi** [comment]
I can recreate the error with the full file, but manually going through the process I dont know, why there is a `'NoneType'` in the loop:

```
>>> parsed.get_payload()
[<email.message.Message object at 0x0000018D6083BE50>, <email.message.Message object at 0x0000018D60838550>, <email.message.Message object at 0x0000018D60838580>]
>>> parsed.get_payload()[0].get('Content-Type')
'text/plain; charset=utf-8'
>>> parsed.get_payload()[1].get('Content-Type') 
'message/delivery-status'
>>> parsed.get_payload()[2].get('Content-Type') 
'message/rfc822'
>>> len(parsed.get_payload())          
3
>>> for b in parsed.get_payload():
...     if b.get('Content-Type')[:b.get('Content-Type').find(";")] == "text/plain":
...             print(b.get('Content-Type'))
... 
text/plain; charset=utf-8
```

**Comment by JM-Lemmi** [comment]
For some reason, there is one more element in `dict_mail['body']` than in `parsed.get_payload()`, which makes no sense?

![image](https://user-images.githubusercontent.com/41118534/141280011-2ace29c4-c7e7-4f87-935f-c0934330e583.png)

**Comment by JM-Lemmi** [comment]
> one more element in dict_mail['body'] than in parsed.get_payload(), which makes no sense?

Ah no. It makes sense, because we recurse over Mailmessage objects actually. Quite a few of them are actually empty, whcih I dont understand...

![image](https://user-images.githubusercontent.com/41118534/141281283-af65700d-a776-4e8c-bcc2-5551c4a6456c.png)

**Comment by JM-Lemmi** [comment]
I propose the following:

- Elements with `Content-Description: Delivery Report` are rendered as Body.
- Elements with `Content-Description: Undelivered Message` are treated as an attachment

**Comment by frereit** [comment]
Sure, that makes sense for now.
I will look into adding an abstraction layer onto this eml parsing because currently we're not very flexible and it's pretty easy to craft an EML that fails to parse.

**Comment by frereit** [comment]
We could also think about just saving the bounce message as an EML file and then generating our own bounce message with just "text/plain" and the original bounce message as an attachment.

We don't expect any other "weird" emls because we can filter any emls that are not just text/plain + attachments with the milter.

**Comment by JM-Lemmi** [comment]
Seems like a better option honestly.

**Comment by frereit** [comment]
Okay, let's keep this issue open for now until we're at least failing gracefully instead of throwing a non descriptive error. I will change the bounce emails to be in the format described above.

**Comment by frereit** [comment]
Closed #61
